### ISSUE TITLE
Incorrect `just_underscores_and_digits` warning for this broken code

### ISSUE THREAD

**Issue by jruderman** [rust-lang/rust-clippy#9724]
### Summary

With this erroneous code, Clippy says "consider choosing a more descriptive name" while pointing at a part of my program that isn't a name.

### Lint Name

just_underscores_and_digits

### Reproducer

I tried running `clippy-driver` on this code:

```rust
fn _id(r: u32 -> u32) {
    r
}

fn main(){}
```

I saw this happen:

```
error: expected parameter name, found `->`
 --> uxn.rs:1:15
  |
1 | fn _id(r: u32 -> u32) {
  |               ^^ expected parameter name

error: expected one of `!`, `(`, `)`, `+`, `,`, `::`, or `<`, found `->`
 --> uxn.rs:1:15
  |
1 | fn _id(r: u32 -> u32) {
  |              -^^ expected one of 7 possible tokens
  |              |
  |              help: missing `,`

warning: consider choosing a more descriptive name
 --> uxn.rs:1:11
  |
1 | fn _id(r: u32 -> u32) {
  |           ^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#just_underscores_and_digits
  = note: `#[warn(clippy::just_underscores_and_digits)]` on by default

error[E0308]: mismatched types
 --> uxn.rs:2:5
  |
1 | fn _id(r: u32 -> u32) {
  |                       - help: try adding a return type: `-> u32`
2 |     r
  |     ^ expected `()`, found `u32`

error: aborting due to 3 previous errors; 1 warning emitted
```

I expected to see this happen:

"warning: consider choosing a more descriptive name" should not appear for this input. (Syntax errors are expected.)

### Version

```text
rustc 1.66.0-nightly (bed4ad65b 2022-10-25)
binary: rustc
commit-hash: bed4ad65bf7a1cef39e3d66b3670189581b3b073
commit-date: 2022-10-25
host: x86_64-apple-darwin
release: 1.66.0-nightly
LLVM version: 15.0.2
```
